libplatsupport/src/mach/eswin/ltimer.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright 2025, UNSW
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <platsupport/io.h>
+#include <platsupport/ltimer.h>
+#include <platsupport/mach/timer.h>
+#include <platsupport/pmem.h>
+#include <platsupport/timer.h>
+#include <utils/io.h>
+#include <utils/util.h>
+
+#include "../../ltimer.h"
+
+enum {
+    COUNTER_TIMER,
+    TIMEOUT_TIMER,
+    NUM_TIMERS
+};
+
+typedef struct {
+    void *vaddr;
+    eswin_timer_t timer[NUM_TIMERS];
+    irq_id_t irq_id[NUM_TIMERS];
+    timer_callback_data_t callback_data;
+    ltimer_callback_fn_t user_callback;
+    void *user_callback_token;
+    ps_io_ops_t ops;
+} eswin_ltimer_t;
+
+static ps_irq_t irqs[] = {
+    {
+        .type = PS_INTERRUPT,
+        .trigger.number = ESWIN_TIMER_IRQ,
+    },
+};
+
+static pmem_region_t pmems[] = {
+    {
+        .type = PMEM_TYPE_DEVICE,
+        .base_addr = ESWIN_TIMER_BASE,
+        .length = ESWIN_TIMER_SIZE,
+    },
+};
+
+#define NUM_IRQS ARRAY_SIZE(irqs)
+#define NUM_PMEMS ARRAY_SIZE(pmems)
+
+static size_t get_num_irqs(void *data)
+{
+    return NUM_IRQS;
+}
+
+static int get_nth_irq(void *data, size_t n, ps_irq_t *irq)
+{
+    assert(n < NUM_IRQS);
+    assert(irq);
+
+    *irq = irqs[n];
+    return 0;
+}
+
+static size_t get_num_pmems(void *data)
+{
+    return NUM_PMEMS;
+}
+
+static int get_nth_pmem(void *data, size_t n, pmem_region_t *paddr)
+{
+    assert(n < NUM_PMEMS);
+    assert(paddr);
+
+    *paddr = pmems[n];
+    return 0;
+}
+
+static int ltimer_handle_irq(void *data, ps_irq_t *irq)
+{
+    assert(data);
+    assert(irq);
+
+    eswin_ltimer_t *timers = (eswin_ltimer_t *)data;
+    long irq_number = irq->irq.number;
+    ltimer_event_t event;
+
+    if (irq_number != irqs[0].irq.number) {
+        ZF_LOGE("Invalid IRQ number %ld received.", irq_number);
+        return EINVAL;
+    }
+
+    if (timers->timer[COUNTER_TIMER].regs->int_status & 0x1) {
+        eswin_timer_handle_irq(&timers->timer[COUNTER_TIMER]);
+        event = LTIMER_OVERFLOW_EVENT;
+    }
+
+    if (timers->user_callback) {
+        timers->user_callback(timers->user_callback_token, event);
+    }
```

Comment:
Shouldn't this be within the first if-check? Otherwise `event` is uninitialised.

Also, if your driver handles timer overflows, from the point of view of the user, there are no timer overflows, so there is no need to tell them this.